test/server/config_validation/xds_fuzz.cc

```diff
@@ -111,32 +110,102 @@ void XdsFuzzTest::close() {
  * @param the listener number to be removed
  * @return the listener as an optional so that it can be used in a delta request
  */
-absl::optional<std::string> XdsFuzzTest::removeListener(uint32_t listener_num) {
-  std::string match = absl::StrCat("listener_", listener_num % ListenersMax);
-
+bool XdsFuzzTest::eraseListener(std::string listener_name) {
   for (auto it = listeners_.begin(); it != listeners_.end(); ++it) {
-    if (it->name() == match) {
+    if (it->name() == listener_name) {
       listeners_.erase(it);
-      return match;
+      return true;
     }
   }
-  return {};
+  return false;
 }
 
 /**
  * remove a route from the list of routes if it exists
  * @param the route number to be removed
  * @return the route as an optional so that it can be used in a delta request
  */
-absl::optional<std::string> XdsFuzzTest::removeRoute(uint32_t route_num) {
-  std::string match = absl::StrCat("route_config_", route_num % RoutesMax);
+bool XdsFuzzTest::eraseRoute(std::string route_name) {
   for (auto it = routes_.begin(); it != routes_.end(); ++it) {
-    if (it->name() == match) {
+    if (it->name() == route_name) {
       routes_.erase(it);
-      return match;
+      return true;
     }
   }
-  return {};
+  return false;
+}
+
+/**
+ * send an xDS response to add a listener and update state accordingly
+ */
+void XdsFuzzTest::addListener(std::string listener_name, std::string route_name) {
+  ENVOY_LOG_MISC(info, "Adding {} with reference to {}", listener_name, route_name);
+  bool removed = eraseListener(listener_name);
+  auto listener = buildListener(listener_name, route_name);
+  listeners_.push_back(listener);
+
+  updateListener(listeners_, {listener}, {});
+  // use waitForAck instead of compareDiscoveryRequest as the client makes
+  // additional discoveryRequests at launch that we might not want to
+  // respond to yet
+  EXPECT_TRUE(waitForAck(Config::TypeUrl::get().Listener, std::to_string(version_)));
+  if (removed) {
+    num_modified_++;
+    test_server_->waitForCounterGe("listener_manager.listener_modified", num_modified_);
```

Comment:
Just to clarify, is this causing example4 to break? Is it because of adding the route the first listener references? (I can't seem to piece that together)

Reply:
What's happening is (I've normalised the listener numbers mod 3 and the routes mod 5 as happens in the code):

1. the first two actions are ignored as there is no active listener yet
2. `listener_1` is added with reference to `route_config_2`, then `route_config_2` is sent to init the server, so `listener_1` is active and `num_active` is 1
3. `listener_1` is added with reference to `route_config_1`, which causes `removed` to be true (as there is already a `listener_1` in `listeners_`, `num_modified_` to be incremented, and `test_server_` to wait for `listeners_modified` to be 1

However, `listeners_modified` does not become 1 until the new `listener_1`, the one that references `route_config_1`, becomes active, i.e. when `route_config_1` is sent, as until then, the original `listener_1` remains active and the new one remains warming.  What should happen is `num_modified` stays at 0 until `route_config_1` is sent, as it is only then that the new `listener_1` takes over from the old one, which is then drained.  I've made these changes which are practically ready for a PR, I'm just cleaning it up a small bit and waiting for this to be merged.  I've factored them into the verifier as some of the logic is more complicated than I initially thought.

Reply:
@htuch do you have any ideas about this? (specifically the **edit**).  Are listeners drained differently when they are removed vs just updated to a new route config?

Reply:
@samflattery there are definitely some interesting corner cases around active/warming clusters and listeners of the same name (I have a fix for some of these at https://github.com/envoyproxy/envoy/pull/12069, but I don't think that affects you right now; it would have been caught hopefully by your fuzzer when you add CDS).

Are there active connections to these listeners? That might influence draining, i.e. why wait to drain if there is no connections.

I'd solve this with code spelunking, I think you're probably going to need to become fairly familiar with listener and listener manager implementations to debug the xDS fuzzer :-D

Reply:
@htuch Read through the PR/issue you linked, that seems to make sense.  Hopefully should be able to catch behaviours like that if I add CDS after this.  As far as I know there are no active connections, I just send the updates but don't ever `makeSingleRequest()` or anything like that, at least for now.  That would make sense, but it's odd that if i just remove a listener instead of updating it, the removed listener seems to drain for the proper length of time (i.e. not just instantaneously like after an update).  I'll do some more digging and see if I can find anything, thanks!